azerothcore/azerothcore-wotlk
# Fix(CORE/loot): players can roll on unique items they already own

Players should not be able to roll on items they already have.
This commit removes that.

Closes https://github.com/azerothcore/azerothcore-wotlk/issues/7896
closes https://github.com/chromiecraft/chromiecraft/issues/1730



## Changes Proposed:
-  Add a check in AllowedForPlayer() if the player already has more than the max allowed of instances of that item
-  

## Issues Addressed:

- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/7896 

## SOURCE:

See issue.

## Tests Performed:

- Tested in strat on magistrate who drops the key to strat 100% and is unique. 
- Worked in master loot (could see but not assign to a player with the key) and group loot (could not see if I had key)


## How to Test the Changes:


1. Find any loot that is unique, unique-equipped or "unique(XX)". For instance key to strat .add 12382
2. Get in a group, add that item to one player
3. Kill a mob that drops that item and see if it can be looted. eg go to strat and clear the first pack, aggro the boss just in front of the church and check. .go 3235.145 -4048.705 108.44 . You can use .die on the mobs and .damage 5000 a couple of times on the boss to get through.

To be checked:
- behaviour on white items
- behaviour on unique-equipepd
- behaviour on unique(XX) 


## Known Issues and TODO List:


- [ ] If nobody can loot the item, you can right-click it but it shows no loot. I recall having this in the past, but am not sure if that was on retail or private, so to be confirmed if that needs improvement.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
